File: Data Structures & Algorithms/valid-tree/submission-2.cpp
class Solution {
public:
    vector<vector<int>> buildAdj(vector<vector<int>>& edges, int n){
        vector<vector<int>> adj(n);
        for(const vector<int> &edge: edges){
            adj[edge[0]].push_back(edge[1]);
            adj[edge[1]].push_back(edge[0]);
        }
        return adj;
    }
    bool dfsCheckCycle(int src, vector<vector<int>>& adj, vector<int> &vis, vector<int> &dfsVis){
        vis[src] = true;
        for(int &neighbour: adj[src]){
            if(dfsVis[neighbour]) return true;
            else if (!vis[neighbour]) if(dfsCheckCycle(neighbour, adj, vis, dfsVis)) return true;
        }
        dfsVis[src] = true;
        return false;
    }
    bool validTree(int n, vector<vector<int>>& edges) {
        if(edges.size()!=n-1) return false;
        // cycle
        vector<vector<int>> adj = buildAdj(edges, n);
        vector<int> vis(n, false);
        vector<int> dfsVis(n, false);
        int component = 0;
        bool ans = true;
        for(int i = 0; i < n; i++){
            if(!vis[i]) {
                ans = !dfsCheckCycle(i, adj, vis, dfsVis);
                component++;
            }
        }
        return ans && (component == 1);
    }
};

File: Data Structures & Algorithms/course-schedule-iv/submission-2.java
class Solution {
     private List<List<Integer>> buildAdjacencyList(int[][] edges, int n, int[] indegree) {
        List<List<Integer>> adj = new ArrayList<>();
        for (int i = 0; i < n; i++) {
            adj.add(new ArrayList<>());
        }
        for (int[] edge : edges) {
            int src = edge[0];
            int dest = edge[1];
            adj.get(src).add(dest);
            indegree[dest]++;
        }
        return adj;
    }
    private boolean bfsPerQuery(int src, int dest, List<List<Integer>> adj, int n){
        boolean[] vis = new boolean[n];
        ArrayDeque<Integer> q = new ArrayDeque<>();
        q.offerLast(src);
        vis[src] = true;
        while(!q.isEmpty()){
            int node = q.pollFirst();
            for(Integer neighbour: adj.get(node)){
                if(neighbour == dest) return true;
                else if(!vis[neighbour]){
                    q.offerLast(neighbour);
                    vis[neighbour] = true;
                }
            }
        }
        return false;
    }
    private void bfs(int src, List<List<Integer>> adj, boolean[][] isReachable, int n, ArrayDeque<Integer> q){
        boolean[] vis = new boolean[n];
        vis[src] = true;
        q.offerLast(src);
        while(!q.isEmpty()){
            int node = q.pollFirst();
            for(Integer neighbour: adj.get(node)){
                if(!vis[neighbour]){
                    q.offerLast(neighbour);
                    vis[neighbour] = true;
                    isReachable[node][neighbour] = true;
                    isReachable[src][neighbour] = true;
                }
            }
        }
        return;
    }
    public List<Boolean> checkIfPrerequisite(int n, int[][] prerequisites, int[][] queries) {
        List<Boolean> ans = new ArrayList<>();
        boolean[][] isReachable = new boolean[n][n];
        int[] indegree = new int[n];
        List<List<Integer>> adj = buildAdjacencyList(prerequisites, n, indegree);
        ArrayDeque<Integer> queue = new ArrayDeque<>();
        for(int i = 0; i < n; i++) if(indegree[i] == 0)  queue.offerLast(i);
        while (!queue.isEmpty()) {
            int u = queue.poll();
            for (Integer v : adj.get(u)) {
                isReachable[u][v] = true;
                for (int i = 0; i < n; i++) {
                    if (isReachable[i][u]) isReachable[i][v] = true;
                }
                if (--indegree[v] == 0) queue.offer(v);
            }
        }
        for (int[] query : queries) {
            ans.add(isReachable[query[0]][query[1]]);
        }
        return ans;
    }
}